version based on git

dalf opened the pull request:
## What does this PR do?

This commit removes:
* the need to update the brand for GIT_URL and GIT_BRANCH.
* these settings:
https://github.com/searxng/searxng/blob/ed0db4d61a2e0dae4953eb813604e2d8834868bb/searx/settings.yml#L13-L16

these values are read from the git repository: it requires the directory to be git repository.

It is possible to call `python -m searx.version freeze` to freeze the current version (it creates the `searx/version_frozen.py` file)
It is useful when the code is install outside git (distro package, docker, etc...)

The code is based on [setuptool_scm](https://github.com/pypa/setuptools_scm/blob/main/src/setuptools_scm/git.py): it calls the `git` command line to retrieve the values. See `searx/version.py`.

## Why is this change important?

* see comments of https://github.com/searxng/searxng/pull/36
* most of the fork don't update the brand ( for example: https://salsa.debian.org/debian/searx/-/blob/master/searx/brand.py https://gitlab.e.foundation/e/cloud/my-spot/-/blob/master/utils/brand.env )
* `make docker.build` : this PR fixes a possible inconsistency in the master branch:
https://github.com/searxng/searxng/blob/ed0db4d61a2e0dae4953eb813604e2d8834868bb/manage#L274-L278
  * GIT_URL is from the brand.
  * SEARX_GIT_VERSION and VERSION_GITCOMMIT are from the local git repository.
  * in a fork, if the brand is not up to date (the user has not changed the git_url) these values won't match.
* in another PR : the bug report form will contains the git URL and commit automatically.

With the change, the footer is always up to date without additional configuration:
![image](https://user-images.githubusercontent.com/1594191/127194119-0c4d377e-6e88-4b1f-ab9b-e4eddc137abc.png)

* `1.0.0-590-f2979c72` comes from `git describe`
* `-dirty` shows that there are some uncommited changes.

## How to test this PR locally?

* `make run`: check the version at the bottom of the page is the right one.
* create a tag locally, check the version is still right.
* call `python -m searx.version freeze`
   * add some change
   * check that version is same as before the changes
* `make docker`:  
   * check the version at the bottom of the page is the right one.
* `make docs`
* use `utils/searx.sh` to install the app
  * check the version
  * check the `utils/searx.sh update` works

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

Closes: https://github.com/searxng/searxng/pull/36

dalf committed:
version based on the git repository

This commit remove the need to update the brand for GIT_URL and GIT_BRANCH:
there are read from the git repository.

It is possible to call python -m searx.version freeze to freeze the current version.
Useful when the code is installed outside git (distro package, docker, etc...)

dalf commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+    return out
+
+
+def get_public_git_url(git_url):
```
Here there is some tricks (?) which works for github and gitlab.

return42 commented on utils/searx.sh:
```diff
@@ -445,16 +442,6 @@ prompt_installation_status(){
-            if ! [ "$instance_setting" = "$branch" ]; then
-                warn_msg "instance brand.git_branch: ${instance_setting}" \
-                         "differs from local clone's branch: ${branch}"
-            fi
```
This part was intended to check if an existing installation is from the same GIT_URL like the _clone_ from which you called the installation procedure.

By example: say you have an existing installation from branch https://github.com/return42/searxng/tree/darmarit.org and at some day you clone from https://github.com/searxng/searxng/tree/master (into your Download folder or where ever) and then you run the installation scripts (or other maintenance tasks) from this fresh clone in your Download folder.

If there comes the day, we have isolated installation procedures from SearXNG repository, we can drop this check.

dalf replied:
I have updated this section:
https://github.com/searxng/searxng/blob/7710c9422fceda1d45cbdda634b80b89ceee2ac6/utils/searx.sh#L444-L458

return42 replied:
I'm testing the installation procedures .. and I will fix it, give me some time (this solution do not work for me).

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+                # os.environ,
+                # try to disable i18n
+                LC_ALL="C",
+                LANGUAGE="",
```
nit: indent by 4 spaces.

dalf replied:
Done (using black on the whole file).

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+    if os.name == "posix" and not isinstance(cmd, (list, tuple)):
+        cmd = shlex.split(cmd)
+
+    p = _popen_pipes(cmd, cwd)
```
IMO we should not implement a function (_popen_pipes) that does nothing more than just one function call / we should place the subprocess.Popen constructor right here.  Having the constructor call in a function, breaks the _with construct_ and is hard to read.

dalf replied:
Done.
Note: in this scenario ( communicate), it seems _with_ advantage is to deal with Ctrl-C:
https://github.com/python/cpython/blob/12fc0d28fcdeca32314d0755d3b30c7d96907440/Lib/subprocess.py#L1014-L1043

return42 replied:
> Note: in this scenario ( communicate), it seems with advantage is to deal with Ctrl-C:

Thanks for the link, but .. Nope, I think the "with" context is to close the stdout & stderr.  The try/except block is around closing stdin .. but its also the first time I looked at this :-)

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+        k: v
+        for k, v in env.items()
+        if not k.startswith("GIT_")
+        or k in ("GIT_EXEC_PATH", "GIT_SSH", "GIT_SSH_COMMAND")
```
Do I miss something, what is the difference of `startswith("GIT_")` and `in ("GIT_EXEC_PATH", "GIT_SSH", "GIT_SSH_COMMAND")`

dalf replied:
`not k.startswith("GIT_") or k in ...` : 
* it accepts `GIT_EXEC_PATH`, `GIT_SSH`, `GIT_SSH_COMMAND` keys
* and all other keys except the keys that start with `GIT_`.

return42 replied:
🤦🏻 .. sorry.

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+        stderr=subprocess.PIPE,
+        cwd=str(cwd),
+        env=dict(
+                no_git_env(os.environ),
```
Except the PATH variable, is there any need for the _no GIT env_?

dalf replied:
See:
https://github.com/searxng/searxng/blob/e5d1cd3c9deccfd2099e48baed855dcc00113ed0/searx/version.py#L37-L41

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+
+
+def get_git_version():
+    git_version = do_ex_safe(r'git describe --match "v[0-9]*\.[0-9]*\.[0-9]*" HEAD')
```
I think we should not use strings to start a command line tool (needs be executed in a shell / or like you do, using `shlex.split(cmd)` ).  For the sake of simplicity I implemented the [fspath.callEXE](https://return42.github.io/fspath/fspath-api/fspath.fspath.html#fspath.fspath.callEXE). 
```python
>>> import fspath
>>> fspath.callEXE('git', 'describe')
('v1.0.0-590-g0686e274\n', '', 0)
```

If you do not want to use fspath (which is ATM only in requirements-dev.txt) you can use:

```python
>>> import subprocess
>>> proc = subprocess.run(['git', 'describe'], stdout=subprocess.PIPE)
>>> proc.stdout.decode()
'v1.0.0-590-g0686e274\n'

```

dalf replied:
> I think we should not use strings to start a command line tool (needs be executed in a shell / or like you do, using shlex.split(cmd) ).

I don't know understand why `shlex.split(cmd)` is a problem ( it is a copy / paste from setup_scm ).
For me, 
`'git describe --match "v[0-9]*\.[0-9]*\.[0-9]*" HEAD'` 
is more readable than
`['git', 'describe', '--match', 'v[0-9]*\\.[0-9]*\\.[0-9]*', 'HEAD']`

return42 replied:
Not really a problem but I would prefer a notation which do not need any kind of *syntax parser".  If you prefer to use one string instead of a argument list you can set `shell=True` in the `subprocess.*` functions (no need for shlex).  Nevertheless ..
[Providing a sequence of arguments is generally preferred, as it allows the module to take care of any required escaping and quoting of arguments (e.g. to permit spaces in file names)](https://docs.python.org/3/library/subprocess.html#frequently-used-arguments)

dalf replied:
* `shell=True` is a different thing: it calls a shell to actually call the command.
* in `version.py`, all commands are hard coded except one.

return42 replied:
> `shell=True` is a different thing: it calls a shell to actually call the command.

Maybe I am used to it differently: for me, when I see a command line in one string I suppose this line will be executed through the shell (including wildcards and other shell builtins) and when I see a list of arguments I suppose that a "program" _is executed_ directly in a subprocess. 

> in `version.py`, all commands are hard coded except one.

I do not understand what you mean by this sentence  .. what means "hard coded" here and "except one" .. all commands in version.py are _hard coded_ (in the meaning of: a constant string is used)

dalf replied:
Is it acceptable to have to string in this particular use case?

> all commands in version.py are hard coded 

yes except:
https://github.com/searxng/searxng/blob/e5d1cd3c9deccfd2099e48baed855dcc00113ed0/searx/version.py#L79

The shell is not involved only `git`, and `origin` comes from a previous git output.

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+    git_version_s = git_version.split('-')
+    git_version_s[0] = git_version_s[0][1:]
+    if len(git_version_s) == 3:
+        git_version_s[2] = git_version_s[2][1:]
```
Juggling with the index in the list is hard to read I recommend to name the values:

```python
tag = proc.stdout.decode().strip()
tag_version, tag_distance, tag_commit = tag.split('-')
```

dalf replied:
Done:
https://github.com/searxng/searxng/blob/e5d1cd3c9deccfd2099e48baed855dcc00113ed0/searx/version.py#L93-L106

Note: If the last commit is a tag `git describe` will return only the tag without distance or commit hash.

return42 commented on setup.py:
```diff
@@ -41,7 +38,7 @@
     author_email='asciimoo@gmail.com',
     license='GNU Affero General Public License',
-    packages=find_packages(exclude=["tests*", "searx_extra"]),
+    packages=find_packages(exclude=["tests*", "searx_extra", "searx.static.themes.simple.node_modules", "searx.static.themes.oscar.node_modules"]),
```
Is this a part of this PR or is this a bugfix out of the scope of this PR?

dalf replied:
Nope, this is a leftover from a test.

dalf replied:
Fixed

return42 reviewed (commented):
@dalf thanks for your explanation / sounds good to me .. but to be true I need some practice to see if there is a trap I don't see right now. Anyway, lets go into this direction ... I made some remarks in my review .. I think the version.py can be improved.

If you want, I can try to improve the implementation .. but I cant promise you when.

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,150 @@
+
+
+class ShellException(Exception):
+
```
Do we need a exception architecture .. why not using the return code of the subprocess directly?

May be it is all better shown by example: I hope I have time this evening .. then I will implement a alternative just to show how I would implement the functionality.

dalf replied:
Look at the `get_git_url_and_branch` function: it raises an exception if any of the git calls fail.

Then
```python
    try:
        GIT_URL, GIT_BRANCH = get_git_url_and_branch()
    except ShellException as e:
        ...
```
is straightforward

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,173 @@
+
+def call_subprocess(cmd, cwd="."):
+    if os.name == "posix" and not isinstance(cmd, (list, tuple)):
+        cmd = shlex.split(cmd)
```
sorry if I'm being annoying :-)  .. What use case (OS) do you have in mind for _non posix_ .. and what happens with the sub process call when ` os.name != "posix"`?

dalf replied:
My take was setup_scm has most probably dealt with various platforms.

But, more in detail, from https://docs.python.org/3/library/subprocess.html#subprocess.Popen :

> On Windows, if args is a sequence, it will be converted to a string in a manner described in Converting an argument sequence to a string on Windows. This is because the underlying CreateProcess() operates on strings.

And according to https://stackoverflow.com/questions/1854/python-what-os-am-i-running-on `os.name != 'posix'` only on Windows (and Jython which is out of scope and perhaps some other implementation I'm not aware of, yes it should be `os.name != 'nt'` according to what I understand even in setup_scm).

But anyway `searx/version.py` works on Windows even with the commit 7710c9422fceda1d45cbdda634b80b89ceee2ac6

return42 commented on searx/version.py:
```diff
@@ -1,26 +1,178 @@
+    if len(sys.argv) >= 2 and sys.argv[1] == "freeze":
+        # freeze the version (to create an archive outside a git repository)
+        # Looks at the current index and checks to see if merges or updates are needed by checking stat() information.
+        call_subprocess_safe("git update-index -q --refresh")
```
For my understanding, for what is this command needed?

Update: it seems that the -q option suppress the return code ... 

```
$ git update-index --refresh
searx/version.py: needs update
error(1)
$ git update-index -q --refresh
(0)
```

dalf replied:
See commit 2d60d80fe7cf04d50933c6e9469b1f625111309d

This is copy / paste from the existing code (from fbe40001d35ea5cf74c92f35d76c642a480a5c9f ):
https://github.com/searxng/searxng/blob/feef577f996b00c4b4b610b6e3a17e2183d20927/manage#L235-L240

There is no reason to use the low level and "git diff --quiet" makes less syscalls (so most probably faster).

dalf replied:
Updated

return42 committed:
[fix] buildenv: first unset environment / then import searx package

The first import of names from the searx package implies loading the
settings.yml.  Before this is done, the enviroment variables must be unset to
not overwrite the values from the settings.yml

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on utils/build_env.py:
```diff
@@ -57,8 +56,9 @@ def _env(*arg, **kwargs):
 
 print('build %s (settings from: %s)' % (brand_env, os.environ['SEARX_SETTINGS_PATH']))
 sys.path.insert(0, repo_root)
-from searx import settings
```
This is the fragile part in the old implementation, read the comments in the lines before.  The problem is, that the settings.yml is imported as soon as you import any searx modules.  If the buildenv should work, you first need to remove any environment (see code above) before importing searx, otherwise the environment will overwrite the values you have modified in the settings.yml.

I fixed it in my branch https://github.com/return42/searxng/commit/6a2d4915a5a8698083d9b21cf6d89c95d2b3caf7

return42 committed:
[fix] searx.sh - get_installed_version_variables()

Fix the quotation marks in the command line to run

    python -m searx.version

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] searx.sh - improve some messages prompted while installation

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 reviewed (approved):
Works flawless / lets merge.

BTW: Thank you for your patience with me :-)
